### PR TITLE
Reorganize singularity engine configuration to reduce plugin binary size

### PR DESCRIPTION
**Pull request by cclerget**
## Description of the Pull Request (PR):

Reorganize singularity engine configuration to get rid of non public struct fields and remove package dependencies which grow plugin binary size uselessly by almost 7 Mbytes.

### This fixes or addresses the following GitHub issues:

 - Related to #4993 

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: internal/app/singularity/plugin_compile_linux.go [hunk 1]
```diff
@@ -223,6 +223,7 @@ func buildPlugin(sourceDir string, bTool buildToolchain) (string, error) {
 		"-o", out,
 		"-mod=readonly",
 		"-trimpath",
+		"-ldflags=-s -w",
```
Comment (dtrudg) [inline internal/app/singularity/plugin_compile_linux.go]: Does `-s` strip-all retain nice things in any Go stack traces? Also - what's the procedure for someone who wants to build a plugin *with* any debugging symbols retained?

File: pkg/runtime/engine/singularity/config/config.go [hunk 2]
```diff
@@ -25,6 +26,21 @@ const (
 	UnderlayLayer = "underlay"
 )
 
+// EngineConfig stores both the JSONConfig and the FileConfig.
```
Comment (dtrudg) [inline pkg/runtime/engine/singularity/config/config.go]: No longer has FileConfig?

File: pkg/runtime/engine/singularity/config/config.go [hunk 3]
```diff
@@ -25,6 +26,21 @@ const (
...
+type EngineConfig struct {
+	JSON      *JSONConfig `json:"jsonConfig"`
+	OciConfig *oci.Config `json:"ociConfig"`
+}
+
+// NewConfig returns singularity.EngineConfig with a parsed FileConfig
+func NewConfig() *EngineConfig {
```
Comment (dtrudg) [inline pkg/runtime/engine/singularity/config/config.go:41]: No longer has Fileconfig?

File: internal/pkg/runtime/engine/singularity/prepare_linux.go [hunk 4]
```diff
@@ -163,6 +163,12 @@ func (e *EngineOperations) PrepareConfig(starterConfig *starter.Config) error {
 		return err
 	}
 
+	// fix for RHEL 6 hosts where ttyname returns a wrong terminal name
+	// (see process_linux.go for details)
+	if e.FileConfig.MountDev == "minimal" || e.EngineConfig.GetContain() {
```
Comment (dtrudg) [inline internal/pkg/runtime/engine/singularity/prepare_linux.go]: If this is a fix, and not just a reorganization, could it be made as a separate PR, so we would be able to confirm the reorganization  maintains exactly the same behaviour before/after?

**Review by dtrudg — COMMENTED** [review]
Hi @cclerget - left a few comments.

Could you say how much of the reduction in size is from the ldflags stripping the output, and how much is from the configuration?

I think it would be nice if...
* reorganization
* ldflags
* devconsole fix

... were in separate commits, so they can be tried in stages to see the effect easily.

File: internal/app/singularity/plugin_compile_linux.go (hunk 1 above)
Comment (cclerget) [reply internal/app/singularity/plugin_compile_linux.go]: Stack trace remains but with less information. There two approaches:
- don't use it by default and let users defined it via `GOFLAGS=-ldflags="-s -w" singularity plugin compile`
- don't use it by default and add a `--strip` flag
- use it by default and add a `--disable-strip` flag

For the second round of plugin the idea is to also generate a basic `Makefile` with `singularity plugin create` to have something like:
- `make`
- `sudo make install`
...
Making the first approach above more appropriate to have something like:
`make debug && sudo make install`

File: internal/pkg/runtime/engine/singularity/prepare_linux.go (hunk 4 above)
Comment (cclerget) [reply internal/pkg/runtime/engine/singularity/prepare_linux.go]: Will create a separate PR. This is a fix because actually the minimal check https://github.com/sylabs/singularity/blob/dbe15bdc62d81cbf6a45504aefefb3beddd25da0/internal/pkg/runtime/engine/singularity/process_linux.go#L71 has no effect as `singularity.conf` can't be parsed from container and the configuration is not passed between stage 1 and 2, so `e.EngineConfig.File.MountDev` is always equal to `yes` in stage 2

File: internal/app/singularity/plugin_compile_linux.go (hunk 1 above)
Comment (dtrudg) [reply internal/app/singularity/plugin_compile_linux.go]: Understood :+1:

**Review by dtrudg — APPROVED** [review]
